oracles/coingecko.js

```diff
@@ -0,0 +1,112 @@
+import assert from 'node:assert';
+import Big from 'big.js';
+import { min } from '../utils.js';
+
+// ~0.01$ pre week(100_800 blocks)
+const PRICE_UPDATE_SPEED = new Big("0.0000001");
+
+//TODO: implement pro-api with api-key
+export class CoinGecko {
+	_url;
+	_cooldown;
+	_updatedAt = 0;
+	_lastPrices = {};
+	_precission;
+
+	_coinGeckoSusdeId;
+	_coinGeckoSusdsId;
+
+	// assets ids used as a keys(hydration's asset ids)
+	_SUSDE;
+	_SUSDS;
+	_USDT;
+	_USDC;
+
+	//Oracle providing prices for sUSDe and sUSDS
+	constructor(config)	 {
+		this._url = config.coinGecko.url;
+		this._cooldown = config.coinGecko.cooldown;
+		this._precission = config.coinGecko.precission;
+
+		this._coinGeckoSusdeId = config.coinGecko.sUSDe;
+		this._coinGeckoSusdsId = config.coinGecko.sUSDS;
+
+		this._SUSDE = config.assets.sUSDe.assetId;
+		this._SUSDS = config.assets.sUSDS.assetId;
+		this._USDT = config.assets.USDT.assetId;
+		this._USDC = config.assets.USDC.assetId;
+
+		this._lastPrices[this._USDT] = new Big("1.0");
+		this._lastPrices[this._USDC] = new Big("1.0");
+		this._lastPrices[this._SUSDS] = new Big(0);
+		this._lastPrices[this._SUSDE] = new Big(0);
+	}
+
+	// Function pulls prices from CoinGecko and initialize oracle's internal state.
+	// This function should be called before oracle is used.
+	// This function doesn't do prices smoothing.
+	async init(currentBlock) {
+		const data = await this._pullOracleData(currentBlock);
+
+		const susdsPrice = new Big(data[this._coinGeckoSusdsId].usd.toFixed(this._precission));
+		const susdePrice = new Big(data[this._coinGeckoSusdeId].usd.toFixed(this._precission));
+
+		this._lastPrices[this._SUSDE] = susdePrice;
+		this._lastPrices[this._SUSDS] = susdsPrice;
+
+		this._updatedAt = currentBlock;
+	}
+
+	//Function pulls prices from coinGecko for sUSDe and sUSDS and updates internal state.
+	//Update happen only after cooldown period.
+	async updateUSDPrices(currentBlock) {
+		if ((currentBlock - this._updatedAt) < this._cooldown) {
+			return;
+		}
+
+		const data = await this._pullOracleData();
+		this._updateOracleData(data, currentBlock);
+	}
+
+	async _pullOracleData() {
+		const res = await fetch(`${this._url}/price?vs_currencies=usd&ids=${this._coinGeckoSusdeId}%2C${this._coinGeckoSusdsId}&precision=${this._precission}`, { method: "GET" });
+		if (!res.ok) {
+			console.error(`ERROR: to pull pricess from CoinGecko. status_code=${res.status}, data=${JSON.Stringify(await res.json())}`)
```

Comment:
i think this would fail during runtime as JSON.Stringify doesnt exist, only with lower letter case.

Also a typo:


```suggestion
			console.error(`ERROR: to pull prices from CoinGecko. status_code=${res.status}, data=${JSON.stringify(await res.json())}`)
```